<a href="https://colab.research.google.com/github/himanshu10095/Credit-Risk-Modelling/blob/main/Credit_Risk_Modelling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Importing necessary Library**

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.metrics import r2_score
from scipy.stats import chi2_contingency
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, precision_recall_fscore_support
import warnings
import os

**Loading the datasets(Excel file)**

In [17]:
from google.colab import files
uploaded = files.upload()

Saving case_study1.xlsx to case_study1 (1).xlsx


In [18]:
import pandas as pd
d1= pd.read_excel("case_study1.xlsx")

In [19]:
d1

PROSPECTID  Total_TL  Tot_Closed_TL  Tot_Active_TL  \
0               1         5              4              1   
1               2         1              0              1   
2               3         8              0              8   
3               4         1              0              1   
4               5         3              2              1   
...           ...       ...            ...            ...   
51331       51332         3              0              3   
51332       51333         4              2              2   
51333       51334         2              1              1   
51334       51335         2              1              1   
51335       51336         1              0              1   

       Total_TL_opened_L6M  Tot_TL_closed_L6M  pct_tl_open_L6M  \
0                        0                  0            0.000   
1                        0                  0            0.000   
2                        1                  0            0.125   
3                        1                  0            1.000   
4                        0                  0            0.000   
...                    ...                ...              ...   
51331                    1                  0            0.333   
51332                    0                  1            0.000   
51333                    1                  1            0.500   
51334                    0                  0            0.000   
51335                    0                  0            0.000   

       pct_tl_closed_L6M  pct_active_tl  pct_closed_tl  ...  CC_TL  \
0                   0.00          0.200          0.800  ...      0   
1                   0.00          1.000          0.000  ...      0   
2                   0.00          1.000          0.000  ...      0   
3                   0.00          1.000          0.000  ...      0   
4                   0.00          0.333          0.667  ...      0   
...                  ...            ...            ...  ...    ...   
51331               0.00          1.000          0.000  ...      0   
51332               0.25          0.500          0.500  ...      0   
51333               0.50          0.500          0.500  ...      0   
51334               0.00          0.500          0.500  ...      0   
51335               0.00          1.000          0.000  ...      0   

       Consumer_TL  Gold_TL  Home_TL  PL_TL  Secured_TL  Unsecured_TL  \
0                0        1        0      4           1             4   
1                1        0        0      0           0             1   
2                6        1        0      0           2             6   
3                0        0        0      0           0             1   
4                0        0        0      0           3             0   
...            ...      ...      ...    ...         ...           ...   
51331            2        0        0      0           0             3   
51332            2        0        0      0           2             2   
51333            2        0        0      0           0             2   
51334            2        0        0      0           0             2   
51335            0        0        0      0           1             0   

       Other_TL  Age_Oldest_TL  Age_Newest_TL  
0             0             72             18  
1             0              7              7  
2             0             47              2  
3             1              5              5  
4             2            131             32  
...         ...            ...            ...  
51331         1             24              5  
51332         0             74              7  
51333         0              9              5  
51334         0             15              8  
51335         0             20             20  

[51336 rows x 26 columns]

In [3]:
from google.colab import files
uploaded = files.upload()

Saving case_study2.xlsx to case_study2.xlsx


In [20]:
import pandas as pd
d2= pd.read_excel("case_study2.xlsx")

Creating the copy of the datasets so that original dataset is unaffected

In [21]:
df1 = d1.copy()
df2 = d2.copy()


**Understanding the dataset**

In [22]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51336 entries, 0 to 51335
Data columns (total 26 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   PROSPECTID            51336 non-null  int64  
 1   Total_TL              51336 non-null  int64  
 2   Tot_Closed_TL         51336 non-null  int64  
 3   Tot_Active_TL         51336 non-null  int64  
 4   Total_TL_opened_L6M   51336 non-null  int64  
 5   Tot_TL_closed_L6M     51336 non-null  int64  
 6   pct_tl_open_L6M       51336 non-null  float64
 7   pct_tl_closed_L6M     51336 non-null  float64
 8   pct_active_tl         51336 non-null  float64
 9   pct_closed_tl         51336 non-null  float64
 10  Total_TL_opened_L12M  51336 non-null  int64  
 11  Tot_TL_closed_L12M    51336 non-null  int64  
 12  pct_tl_open_L12M      51336 non-null  float64
 13  pct_tl_closed_L12M    51336 non-null  float64
 14  Tot_Missed_Pmnt       51336 non-null  int64  
 15  Auto_TL            

In [23]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51336 entries, 0 to 51335
Data columns (total 62 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   PROSPECTID                    51336 non-null  int64  
 1   time_since_recent_payment     51336 non-null  int64  
 2   time_since_first_deliquency   51336 non-null  int64  
 3   time_since_recent_deliquency  51336 non-null  int64  
 4   num_times_delinquent          51336 non-null  int64  
 5   max_delinquency_level         51336 non-null  int64  
 6   max_recent_level_of_deliq     51336 non-null  int64  
 7   num_deliq_6mts                51336 non-null  int64  
 8   num_deliq_12mts               51336 non-null  int64  
 9   num_deliq_6_12mts             51336 non-null  int64  
 10  max_deliq_6mts                51336 non-null  int64  
 11  max_deliq_12mts               51336 non-null  int64  
 12  num_times_30p_dpd             51336 non-null  int64  
 13  n

In [24]:
df1.describe()

PROSPECTID      Total_TL  Tot_Closed_TL  Tot_Active_TL  \
count  51336.000000  51336.000000   51336.000000   51336.000000   
mean   25668.500000      4.858598       2.770415       2.088184   
std    14819.571046      7.177116       5.941680       2.290774   
min        1.000000      1.000000       0.000000       0.000000   
25%    12834.750000      1.000000       0.000000       1.000000   
50%    25668.500000      2.000000       1.000000       1.000000   
75%    38502.250000      5.000000       3.000000       3.000000   
max    51336.000000    235.000000     216.000000      47.000000   

       Total_TL_opened_L6M  Tot_TL_closed_L6M  pct_tl_open_L6M  \
count         51336.000000       51336.000000     51336.000000   
mean              0.736754           0.428919         0.184574   
std               1.296717           0.989972         0.297414   
min               0.000000           0.000000         0.000000   
25%               0.000000           0.000000         0.000000   
50%               0.000000           0.000000         0.000000   
75%               1.000000           1.000000         0.308000   
max              27.000000          19.000000         1.000000   

       pct_tl_closed_L6M  pct_active_tl  pct_closed_tl  ...         CC_TL  \
count       51336.000000   51336.000000   51336.000000  ...  51336.000000   
mean            0.089095       0.577542       0.422458  ...      0.124981   
std             0.205635       0.379867       0.379867  ...      0.505201   
min             0.000000       0.000000       0.000000  ...      0.000000   
25%             0.000000       0.250000       0.000000  ...      0.000000   
50%             0.000000       0.556000       0.444000  ...      0.000000   
75%             0.053000       1.000000       0.750000  ...      0.000000   
max             1.000000       1.000000       1.000000  ...     27.000000   

        Consumer_TL       Gold_TL       Home_TL         PL_TL    Secured_TL  \
count  51336.000000  51336.000000  51336.000000  51336.000000  51336.000000   
mean       1.136084      1.561847      0.070146      0.282511      2.844904   
std        2.227997      5.376434      0.340861      0.858168      6.187177   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      1.000000   
75%        1.000000      1.000000      0.000000      0.000000      3.000000   
max       41.000000    235.000000     10.000000     29.000000    235.000000   

       Unsecured_TL      Other_TL  Age_Oldest_TL  Age_Newest_TL  
count  51336.000000  51336.000000   51336.000000   51336.000000  
mean       2.013694      1.089762     -32.575639     -62.149525  
std        3.198322      2.417496    2791.869609    2790.818622  
min        0.000000      0.000000  -99999.000000  -99999.000000  
25%        0.000000      0.000000      14.000000       4.000000  
50%        1.000000      0.000000      33.000000       8.000000  
75%        2.000000      1.000000      64.000000      17.000000  
max       55.000000     80.000000     392.000000     392.000000  

[8 rows x 26 columns]

In [25]:
df2.describe()

PROSPECTID  time_since_recent_payment  time_since_first_deliquency  \
count  51336.000000               51336.000000                 51336.000000   
mean   25668.500000               -8129.961314                -70020.091320   
std    14819.571046               27749.328514                 45823.312757   
min        1.000000              -99999.000000                -99999.000000   
25%    12834.750000                  46.000000                -99999.000000   
50%    25668.500000                  70.000000                -99999.000000   
75%    38502.250000                 161.000000                     8.000000   
max    51336.000000                6065.000000                    35.000000   

       time_since_recent_deliquency  num_times_delinquent  \
count                  51336.000000          51336.000000   
mean                  -70022.375838              1.573749   
std                    45819.820741              4.165012   
min                   -99999.000000              0.000000   
25%                   -99999.000000              0.000000   
50%                   -99999.000000              0.000000   
75%                        3.000000              1.000000   
max                       35.000000             74.000000   

       max_delinquency_level  max_recent_level_of_deliq  num_deliq_6mts  \
count           51336.000000               51336.000000    51336.000000   
mean           -70003.987085                  13.521953        0.184977   
std             45847.976100                  53.336976        0.710240   
min            -99999.000000                   0.000000        0.000000   
25%            -99999.000000                   0.000000        0.000000   
50%            -99999.000000                   0.000000        0.000000   
75%                15.000000                  10.000000        0.000000   
max               900.000000                 900.000000       12.000000   

       num_deliq_12mts  num_deliq_6_12mts  ...  PL_utilization       PL_Flag  \
count     51336.000000       51336.000000  ...    51336.000000  51336.000000   
mean          0.480053           0.295076  ...   -86556.225194      0.167874   
std           1.522210           1.027471  ...    34111.414750      0.373758   
min           0.000000           0.000000  ...   -99999.000000      0.000000   
25%           0.000000           0.000000  ...   -99999.000000      0.000000   
50%           0.000000           0.000000  ...   -99999.000000      0.000000   
75%           0.000000           0.000000  ...   -99999.000000      0.000000   
max          28.000000          20.000000  ...        1.708000      1.000000   

       pct_PL_enq_L6m_of_L12m  pct_CC_enq_L6m_of_L12m  pct_PL_enq_L6m_of_ever  \
count            51336.000000            51336.000000            51336.000000   
mean                 0.190414                0.065182                0.170492   
std                  0.376218                0.235706                0.350209   
min                  0.000000                0.000000                0.000000   
25%                  0.000000                0.000000                0.000000   
50%                  0.000000                0.000000                0.000000   
75%                  0.000000                0.000000                0.000000   
max                  1.000000                1.000000                1.000000   

       pct_CC_enq_L6m_of_ever  max_unsec_exposure_inPct       HL_Flag  \
count            51336.000000              51336.000000  51336.000000   
mean                 0.056302             -45127.943635      0.271116   
std                  0.213506              49795.784556      0.444540   
min                  0.000000             -99999.000000      0.000000   
25%                  0.000000             -99999.000000      0.000000   
50%                  0.000000                  0.333000      0.000000   
75%                  0.000000                  2.164250      1.000000   
max                  1.000000       

In [26]:
df1.shape

(51336, 26)

In [27]:
df2.shape

(51336, 62)

Removing the rows consisting Null values(-99999) from df1

In [28]:
df1 = df1.loc[df1['Age_Oldest_TL'] != -99999]

In [29]:
df1.shape

(51296, 26)

As df2 contain many Null values(-99999) so if a column contain more than 10000 null values than we will remove that column

In [30]:
columns_to_be_removed = []

for i in df2.columns:
    if df2.loc[df2[i] == -99999].shape[0] > 10000:
        columns_to_be_removed .append(i)


List of columns to be removed

In [31]:
columns_to_be_removed


['time_since_first_deliquency',
 'time_since_recent_deliquency',
 'max_delinquency_level',
 'max_deliq_6mts',
 'max_deliq_12mts',
 'CC_utilization',
 'PL_utilization',
 'max_unsec_exposure_inPct']

In [32]:
df2 = df2.drop(columns_to_be_removed, axis =1)


In [33]:
for i in df2.columns:
    df2 = df2.loc[ df2[i] != -99999 ]


In [34]:
df2.shape

(42066, 54)

Merging the two datsets into a one single dataset with the help of common column

In [35]:
for i in list(df1.columns):
    if i in list(df2.columns):
        print (i)

PROSPECTID


In [36]:
df = pd. merge ( df1, df2, how ='inner', left_on = ['PROSPECTID'], right_on = ['PROSPECTID'] )

In [37]:
df.describe()

PROSPECTID      Total_TL  Tot_Closed_TL  Tot_Active_TL  \
count  42064.000000  42064.000000   42064.000000   42064.000000   
mean   25649.827477      5.262980       2.967383       2.295597   
std    14844.173396      7.463383       6.141098       2.404086   
min        1.000000      1.000000       0.000000       0.000000   
25%    12776.750000      1.000000       0.000000       1.000000   
50%    25706.500000      3.000000       1.000000       2.000000   
75%    38518.250000      6.000000       3.000000       3.000000   
max    51336.000000    235.000000     216.000000      47.000000   

       Total_TL_opened_L6M  Tot_TL_closed_L6M  pct_tl_open_L6M  \
count         42064.000000        42064.00000     42064.000000   
mean              0.812643            0.48992         0.179032   
std               1.383559            1.05892         0.278043   
min               0.000000            0.00000         0.000000   
25%               0.000000            0.00000         0.000000   
50%               0.000000            0.00000         0.000000   
75%               1.000000            1.00000         0.333000   
max              27.000000           19.00000         1.000000   

       pct_tl_closed_L6M  pct_active_tl  pct_closed_tl  ...  \
count       42064.000000   42064.000000   42064.000000  ...   
mean            0.097783       0.577452       0.422548  ...   
std             0.210957       0.366110       0.366110  ...   
min             0.000000       0.000000       0.000000  ...   
25%             0.000000       0.286000       0.000000  ...   
50%             0.000000       0.545000       0.455000  ...   
75%             0.100000       1.000000       0.714000  ...   
max             1.000000       1.000000       1.000000  ...   

       pct_currentBal_all_TL       CC_Flag       PL_Flag  \
count           42064.000000  42064.000000  42064.000000   
mean                0.883693      0.102962      0.193063   
std                40.622275      0.303913      0.394707   
min                 0.000000      0.000000      0.000000   
25%                 0.152000      0.000000      0.000000   
50%                 0.600000      0.000000      0.000000   
75%                 0.860000      0.000000      0.000000   
max              6327.500000      1.000000      1.000000   

       pct_PL_enq_L6m_of_L12m  pct_CC_enq_L6m_of_L12m  pct_PL_enq_L6m_of_ever  \
count            42064.000000            42064.000000            42064.000000   
mean                 0.219169                0.074833                0.195497   
std                  0.395100                0.250658                0.367414   
min                  0.000000                0.000000                0.000000   
25%                  0.000000                0.000000                0.000000   
50%                  0.000000                0.000000                0.000000   
75%                  0.000000                0.000000                0.000000   
max                  1.000000                1.000000                1.000000   

       pct_CC_enq_L6m_of_ever       HL_Flag       GL_Flag  Credit_Score  
count            42064.000000  42064.000000  42064.000000  42064.000000  
mean                 0.064186      0.252235      0.056580    679.326336  
std                  0.225989      0.434300      0.231042     21.133619  
min                  0.000000      0.000000      0.000000    469.000000  
25%                  0.000000      0.000000      0.000000    668.000000  
50%                  0.000000      0.000000      0.000000    679.000000  
75%                  0.000000      1.000000      0.000000    690.000000  
max                  1.000000      1.000000      1.000000    809.000000  

[8 rows x 73 columns]

New Merged Datset is created

In [38]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42064 entries, 0 to 42063
Data columns (total 79 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   PROSPECTID                  42064 non-null  int64  
 1   Total_TL                    42064 non-null  int64  
 2   Tot_Closed_TL               42064 non-null  int64  
 3   Tot_Active_TL               42064 non-null  int64  
 4   Total_TL_opened_L6M         42064 non-null  int64  
 5   Tot_TL_closed_L6M           42064 non-null  int64  
 6   pct_tl_open_L6M             42064 non-null  float64
 7   pct_tl_closed_L6M           42064 non-null  float64
 8   pct_active_tl               42064 non-null  float64
 9   pct_closed_tl               42064 non-null  float64
 10  Total_TL_opened_L12M        42064 non-null  int64  
 11  Tot_TL_closed_L12M          42064 non-null  int64  
 12  pct_tl_open_L12M            42064 non-null  float64
 13  pct_tl_closed_L12M          420

**Feature Engineering**--
We will convert all the categorical columns into numerical columns so that we can make our machine learning model

Analyzing Categorical columns

In [39]:
for i in df.columns:
    if df[i].dtype == 'object':
        print(i)



MARITALSTATUS
EDUCATION
GENDER
last_prod_enq2
first_prod_enq2
Approved_Flag


In [40]:
print(df['MARITALSTATUS'].value_counts())
print(" ")
print(df['EDUCATION'].value_counts())
print(" ")
print(df['GENDER'].value_counts())
print(" ")
print(df['last_prod_enq2'].value_counts())
print(" ")
print(df['first_prod_enq2'].value_counts())
print(" ")
print(df['Approved_Flag'].value_counts())

MARITALSTATUS
Married    30886
Single     11178
Name: count, dtype: int64
 
EDUCATION
GRADUATE          14140
12TH              11703
SSC                7241
UNDER GRADUATE     4572
OTHERS             2291
POST-GRADUATE      1898
PROFESSIONAL        219
Name: count, dtype: int64
 
GENDER
M    37345
F     4719
Name: count, dtype: int64
 
last_prod_enq2
ConsumerLoan    16480
others          13653
PL               7553
CC               2195
AL               1353
HL                830
Name: count, dtype: int64
 
first_prod_enq2
others          20640
ConsumerLoan    11075
PL               4431
AL               2641
CC               1988
HL               1289
Name: count, dtype: int64
 
Approved_Flag
P2    25452
P3     6440
P4     5264
P1     4908
Name: count, dtype: int64


Applying **"Chi-Square test"** to check the assosiation of all the categorical columns with our approved flag column(Result column)

In [41]:
for i in ['MARITALSTATUS', 'EDUCATION', 'GENDER', 'last_prod_enq2', 'first_prod_enq2']:
    chi2, pval, _, _ = chi2_contingency(pd.crosstab(df[i], df['Approved_Flag']))
    print(i, '---', pval)

MARITALSTATUS --- 3.578180861038862e-233
EDUCATION --- 2.6942265249737532e-30
GENDER --- 1.907936100186563e-05
last_prod_enq2 --- 0.0
first_prod_enq2 --- 7.84997610555419e-287


As the pval for all the columns is less than 0.5 so their is assosiation between each column and approved flag

In [42]:
numeric_columns =[]
for i in df.columns:
  if df[i].dtype != 'object' and i not in ['PROSPECTID','Approved_Flag']:
    numeric_columns.append(i)


print(numeric_columns)


['Total_TL', 'Tot_Closed_TL', 'Tot_Active_TL', 'Total_TL_opened_L6M', 'Tot_TL_closed_L6M', 'pct_tl_open_L6M', 'pct_tl_closed_L6M', 'pct_active_tl', 'pct_closed_tl', 'Total_TL_opened_L12M', 'Tot_TL_closed_L12M', 'pct_tl_open_L12M', 'pct_tl_closed_L12M', 'Tot_Missed_Pmnt', 'Auto_TL', 'CC_TL', 'Consumer_TL', 'Gold_TL', 'Home_TL', 'PL_TL', 'Secured_TL', 'Unsecured_TL', 'Other_TL', 'Age_Oldest_TL', 'Age_Newest_TL', 'time_since_recent_payment', 'num_times_delinquent', 'max_recent_level_of_deliq', 'num_deliq_6mts', 'num_deliq_12mts', 'num_deliq_6_12mts', 'num_times_30p_dpd', 'num_times_60p_dpd', 'num_std', 'num_std_6mts', 'num_std_12mts', 'num_sub', 'num_sub_6mts', 'num_sub_12mts', 'num_dbt', 'num_dbt_6mts', 'num_dbt_12mts', 'num_lss', 'num_lss_6mts', 'num_lss_12mts', 'recent_level_of_deliq', 'tot_enq', 'CC_enq', 'CC_enq_L6m', 'CC_enq_L12m', 'PL_enq', 'PL_enq_L6m', 'PL_enq_L12m', 'time_since_recent_enq', 'enq_L12m', 'enq_L6m', 'enq_L3m', 'AGE', 'NETMONTHLYINCOME', 'Time_With_Curr_Empr', 'pct_

Variance Inflation Factor (VIF) will help us  to detect multicollinearity in multiple our  model

In [43]:
vif_data = df[numeric_columns]
total_columns = vif_data.shape[1]
columns_to_be_kept = []
column_index = 0



for i in range (0,total_columns):

    vif_value = variance_inflation_factor(vif_data, column_index)
    print (column_index,'---',vif_value)


    if vif_value <= 6:
        columns_to_be_kept.append( numeric_columns[i] )
        column_index = column_index+1

    else:
        vif_data = vif_data.drop([ numeric_columns[i] ] , axis=1)

/usr/local/lib/python3.11/dist-packages/statsmodels/stats/outliers_influence.py:197: RuntimeWarning: divide by zero encountered in scalar divide
  vif = 1. / (1. - r_squared_i)


0 --- inf


/usr/local/lib/python3.11/dist-packages/statsmodels/stats/outliers_influence.py:197: RuntimeWarning: divide by zero encountered in scalar divide
  vif = 1. / (1. - r_squared_i)


0 --- inf
0 --- 11.320180023967996
0 --- 8.363698035000336
0 --- 6.520647877790928
0 --- 5.149501618212625
1 --- 2.611111040579735


/usr/local/lib/python3.11/dist-packages/statsmodels/stats/outliers_influence.py:197: RuntimeWarning: divide by zero encountered in scalar divide
  vif = 1. / (1. - r_squared_i)


2 --- inf
2 --- 1788.7926256209232
2 --- 8.601028256477228
2 --- 3.832800792153077
3 --- 6.099653381646723
3 --- 5.581352009642766
4 --- 1.985584353098778


/usr/local/lib/python3.11/dist-packages/statsmodels/stats/outliers_influence.py:197: RuntimeWarning: divide by zero encountered in scalar divide
  vif = 1. / (1. - r_squared_i)


5 --- inf
5 --- 4.80953830281934
6 --- 23.270628983464636
6 --- 30.595522588100053
6 --- 4.384346405965583
7 --- 3.0646584155234238
8 --- 2.898639771299251
9 --- 4.377876915347324
10 --- 2.207853583695844
11 --- 4.916914200506864
12 --- 5.214702030064725
13 --- 3.3861625024231476
14 --- 7.840583309478997
14 --- 5.255034641721434


/usr/local/lib/python3.11/dist-packages/statsmodels/stats/outliers_influence.py:197: RuntimeWarning: divide by zero encountered in scalar divide
  vif = 1. / (1. - r_squared_i)


15 --- inf
15 --- 7.380634506427238
15 --- 1.4210050015175733
16 --- 8.083255010190316
16 --- 1.6241227524040114
17 --- 7.257811920140003
17 --- 15.59624383268298
17 --- 1.825857047132431
18 --- 1.5080839450032664
19 --- 2.172088834824578
20 --- 2.6233975535272274
21 --- 2.2959970812106176
22 --- 7.360578319196446
22 --- 2.1602387773102567
23 --- 2.8686288267891467
24 --- 6.458218003637272
24 --- 2.8474118865638247
25 --- 4.753198156284083
26 --- 16.22735475594825
26 --- 6.424377256363877
26 --- 8.887080381808678
26 --- 2.3804746142952653
27 --- 8.60951347651454
27 --- 13.06755093547673
27 --- 3.500040056654653
28 --- 1.9087955874813773
29 --- 17.006562234161628
29 --- 10.730485153719197
29 --- 2.3538497522950275
30 --- 22.10485591513649
30 --- 2.7971639638512924
31 --- 3.424171203217696
32 --- 10.175021454450922
32 --- 6.408710354561292
32 --- 1.001151196262563
33 --- 3.069197305397273
34 --- 2.8091261600643724
35 --- 20.249538381980678
35 --- 15.864576541593774
35 --- 1.8331649740532

In [44]:
from scipy.stats import f_oneway

columns_to_be_kept_numerical = []

for i in columns_to_be_kept:
    a = list(df[i])
    b = list(df['Approved_Flag'])

    group_P1 = [value for value, group in zip(a, b) if group == 'P1']
    group_P2 = [value for value, group in zip(a, b) if group == 'P2']
    group_P3 = [value for value, group in zip(a, b) if group == 'P3']
    group_P4 = [value for value, group in zip(a, b) if group == 'P4']


    f_statistic, p_value = f_oneway(group_P1, group_P2, group_P3, group_P4)

    if p_value <= 0.05:
        columns_to_be_kept_numerical.append(i)


In [45]:
features = columns_to_be_kept_numerical + ['MARITALSTATUS', 'EDUCATION', 'GENDER', 'last_prod_enq2', 'first_prod_enq2']
df = df[features + ['Approved_Flag']]

In [46]:
['MARITALSTATUS', 'EDUCATION', 'GENDER' , 'last_prod_enq2' ,'first_prod_enq2']

df['MARITALSTATUS'].unique()
df['EDUCATION'].unique()
df['GENDER'].unique()
df['last_prod_enq2'].unique()
df['first_prod_enq2'].unique()


array(['PL', 'ConsumerLoan', 'others', 'AL', 'HL', 'CC'], dtype=object)

Assigning the numerical value to our categorical columns (encoding)

In [47]:
df.loc[df['EDUCATION'] == 'SSC',['EDUCATION']]              = 1
df.loc[df['EDUCATION'] == '12TH',['EDUCATION']]             = 2
df.loc[df['EDUCATION'] == 'GRADUATE',['EDUCATION']]         = 3
df.loc[df['EDUCATION'] == 'UNDER GRADUATE',['EDUCATION']]   = 3
df.loc[df['EDUCATION'] == 'POST-GRADUATE',['EDUCATION']]    = 4
df.loc[df['EDUCATION'] == 'OTHERS',['EDUCATION']]           = 1
df.loc[df['EDUCATION'] == 'PROFESSIONAL',['EDUCATION']]     = 3

In [48]:
df['EDUCATION'].value_counts()

EDUCATION
3    18931
2    11703
1     9532
4     1898
Name: count, dtype: int64

In [49]:
df['EDUCATION'] = df['EDUCATION'].astype(int)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42064 entries, 0 to 42063
Data columns (total 43 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   pct_tl_open_L6M            42064 non-null  float64
 1   pct_tl_closed_L6M          42064 non-null  float64
 2   Tot_TL_closed_L12M         42064 non-null  int64  
 3   pct_tl_closed_L12M         42064 non-null  float64
 4   Tot_Missed_Pmnt            42064 non-null  int64  
 5   CC_TL                      42064 non-null  int64  
 6   Home_TL                    42064 non-null  int64  
 7   PL_TL                      42064 non-null  int64  
 8   Secured_TL                 42064 non-null  int64  
 9   Unsecured_TL               42064 non-null  int64  
 10  Other_TL                   42064 non-null  int64  
 11  Age_Oldest_TL              42064 non-null  int64  
 12  Age_Newest_TL              42064 non-null  int64  
 13  time_since_recent_payment  42064 non-null  int

In [50]:
df_encoded = pd.get_dummies(df, columns=['MARITALSTATUS','GENDER', 'last_prod_enq2' ,'first_prod_enq2'])

df_encoded.info()
k = df_encoded.describe()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42064 entries, 0 to 42063
Data columns (total 55 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   pct_tl_open_L6M               42064 non-null  float64
 1   pct_tl_closed_L6M             42064 non-null  float64
 2   Tot_TL_closed_L12M            42064 non-null  int64  
 3   pct_tl_closed_L12M            42064 non-null  float64
 4   Tot_Missed_Pmnt               42064 non-null  int64  
 5   CC_TL                         42064 non-null  int64  
 6   Home_TL                       42064 non-null  int64  
 7   PL_TL                         42064 non-null  int64  
 8   Secured_TL                    42064 non-null  int64  
 9   Unsecured_TL                  42064 non-null  int64  
 10  Other_TL                      42064 non-null  int64  
 11  Age_Oldest_TL                 42064 non-null  int64  
 12  Age_Newest_TL                 42064 non-null  int64  
 13  t

Feature which are required for our model are selected now we can make our model

**Model Building**

Random Forest model

In [51]:
y = df_encoded['Approved_Flag']
x = df_encoded. drop ( ['Approved_Flag'], axis = 1 )


x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)
rf_classifier = RandomForestClassifier(n_estimators = 200, random_state=42)
rf_classifier.fit(x_train, y_train)
y_pred = rf_classifier.predict(x_test)



accuracy = accuracy_score(y_test, y_pred)
print ()
print(f'Accuracy: {accuracy}')
print ()
precision, recall, f1_score, _ = precision_recall_fscore_support(y_test, y_pred)


for i, v in enumerate(['p1', 'p2', 'p3', 'p4']):
    print(f"Class {v}:")
    print(f"Precision: {precision[i]}")
    print(f"Recall: {recall[i]}")
    print(f"F1 Score: {f1_score[i]}")
    print()


Accuracy: 0.7636990372043266

Class p1:
Precision: 0.8370457209847597
Recall: 0.7041420118343196
F1 Score: 0.7648634172469202

Class p2:
Precision: 0.7957519116397621
Recall: 0.9282457879088206
F1 Score: 0.856907593778591

Class p3:
Precision: 0.4423380726698262
Recall: 0.21132075471698114
F1 Score: 0.28600612870275793

Class p4:
Precision: 0.7178502879078695
Recall: 0.7269193391642371
F1 Score: 0.7223563495895703



Xg-Boost model

In [52]:
import xgboost as xgb
from sklearn.preprocessing import LabelEncoder

xgb_classifier = xgb.XGBClassifier(objective='multi:softmax',  num_class=4)



y = df_encoded['Approved_Flag']
x = df_encoded. drop ( ['Approved_Flag'], axis = 1 )


label_encoder = LabelEncoder()
y_encoded = label_encoder.fit_transform(y)


x_train, x_test, y_train, y_test = train_test_split(x, y_encoded, test_size=0.2, random_state=42)


xgb_classifier.fit(x_train, y_train)
y_pred = xgb_classifier.predict(x_test)

accuracy = accuracy_score(y_test, y_pred)
print ()
print(f'Accuracy: {accuracy:.2f}')
print ()

precision, recall, f1_score, _ = precision_recall_fscore_support(y_test, y_pred)

for i, v in enumerate(['p1', 'p2', 'p3', 'p4']):
    print(f"Class {v}:")
    print(f"Precision: {precision[i]}")
    print(f"Recall: {recall[i]}")
    print(f"F1 Score: {f1_score[i]}")
    print()


Accuracy: 0.78

Class p1:
Precision: 0.823906083244397
Recall: 0.7613412228796844
F1 Score: 0.7913890312660175

Class p2:
Precision: 0.8255418233924413
Recall: 0.913577799801784
F1 Score: 0.8673315769665035

Class p3:
Precision: 0.4756380510440835
Recall: 0.30943396226415093
F1 Score: 0.37494284407864653

Class p4:
Precision: 0.7342386032977691
Recall: 0.7356656948493683
F1 Score: 0.7349514563106796



Decision Tree Model

In [53]:
from sklearn.tree import DecisionTreeClassifier

y = df_encoded['Approved_Flag']
x = df_encoded. drop ( ['Approved_Flag'], axis = 1 )

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

dt_model = DecisionTreeClassifier(max_depth=20, min_samples_split=10)
dt_model.fit(x_train, y_train)
y_pred = dt_model.predict(x_test)

accuracy = accuracy_score(y_test, y_pred)
print ()
print(f"Accuracy: {accuracy:.2f}")
print ()

precision, recall, f1_score, _ = precision_recall_fscore_support(y_test, y_pred)

for i, v in enumerate(['p1', 'p2', 'p3', 'p4']):
    print(f"Class {v}:")
    print(f"Precision: {precision[i]}")
    print(f"Recall: {recall[i]}")
    print(f"F1 Score: {f1_score[i]}")
    print()



Accuracy: 0.71

Class p1:
Precision: 0.7252964426877471
Recall: 0.7238658777120316
F1 Score: 0.7245804540967423

Class p2:
Precision: 0.8093944099378882
Recall: 0.8265609514370664
F1 Score: 0.8178876140041189

Class p3:
Precision: 0.34394904458598724
Recall: 0.3260377358490566
F1 Score: 0.33475397132894225

Class p4:
Precision: 0.6515609264853978
Recall: 0.6287657920310982
F1 Score: 0.6399604352126608



Doing scaling of data

In [54]:
from sklearn.preprocessing import StandardScaler

columns_to_be_scaled = ['Age_Oldest_TL','Age_Newest_TL','time_since_recent_payment',
'max_recent_level_of_deliq','recent_level_of_deliq',
'time_since_recent_enq','NETMONTHLYINCOME','Time_With_Curr_Empr']

for i in columns_to_be_scaled:
    column_data = df_encoded[i].values.reshape(-1, 1)
    scaler = StandardScaler()
    scaled_column = scaler.fit_transform(column_data)
    df_encoded[i] = scaled_column



In [55]:
import xgboost as xgb
from sklearn.preprocessing import LabelEncoder

xgb_classifier = xgb.XGBClassifier(objective='multi:softmax',  num_class=4)



y = df_encoded['Approved_Flag']
x = df_encoded. drop ( ['Approved_Flag'], axis = 1 )


label_encoder = LabelEncoder()
y_encoded = label_encoder.fit_transform(y)


x_train, x_test, y_train, y_test = train_test_split(x, y_encoded, test_size=0.2, random_state=42)

In [56]:
xgb_classifier.fit(x_train, y_train)
y_pred = xgb_classifier.predict(x_test)

accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')


precision, recall, f1_score, _ = precision_recall_fscore_support(y_test, y_pred)

for i, v in enumerate(['p1', 'p2', 'p3', 'p4']):
    print(f"Class {v}:")
    print(f"Precision: {precision[i]}")
    print(f"Recall: {recall[i]}")
    print(f"F1 Score: {f1_score[i]}")
    print()

Accuracy: 0.78
Class p1:
Precision: 0.823906083244397
Recall: 0.7613412228796844
F1 Score: 0.7913890312660175

Class p2:
Precision: 0.8255418233924413
Recall: 0.913577799801784
F1 Score: 0.8673315769665035

Class p3:
Precision: 0.4756380510440835
Recall: 0.30943396226415093
F1 Score: 0.37494284407864653

Class p4:
Precision: 0.7342386032977691
Recall: 0.7356656948493683
F1 Score: 0.7349514563106796



After applying standard sacling we can see no improvement in metrices

Now , we furthet procees with Hyperparameter tuning in xgboost as it provide highest accuracy among the three

In [57]:
from sklearn.model_selection import GridSearchCV
x_train, x_test, y_train, y_test = train_test_split(x, y_encoded, test_size=0.2, random_state=42)

# Define the XGBClassifier with the initial set of hyperparameters
xgb_model = xgb.XGBClassifier(objective='multi:softmax', num_class=4)

# Define the parameter grid for hyperparameter tuning

param_grid = {
    'n_estimators': [50, 100, 200],
    'max_depth': [3, 5, 7],
    'learning_rate': [0.01, 0.1, 0.2],
}

grid_search = GridSearchCV(estimator=xgb_model, param_grid=param_grid, cv=3, scoring='accuracy', n_jobs=-1)
grid_search.fit(x_train, y_train)

GridSearchCV(cv=3,
             estimator=XGBClassifier(base_score=None, booster=None,
                                     callbacks=None, colsample_bylevel=None,
                                     colsample_bynode=None,
                                     colsample_bytree=None, device=None,
                                     early_stopping_rounds=None,
                                     enable_categorical=False, eval_metric=None,
                                     feature_types=None, gamma=None,
                                     grow_policy=None, importance_type=None,
                                     interaction_constraints=None,
                                     learning_rate=None,...
                                     max_cat_threshold=None,
                                     max_cat_to_onehot=None,
                                     max_delta_step=None, max_depth=None,
                                     max_leaves=None, min_child_weight=None,
                                     missing=nan, monotone_constraints=None,
                                     multi_strategy=None, n_estimators=None,
                                     n_jobs=None, num_class=4,
                                     num_parallel_tree=None, ...),
             n_jobs=-1,
             param_grid={'learning_rate': [0.01, 0.1, 0.2],
                         'max_depth': [3, 5, 7],
                         'n_estimators': [50, 100, 200]},
             scoring='accuracy')

In [58]:
print("Best Hyperparameters:", grid_search.best_params_)

Best Hyperparameters: {'learning_rate': 0.2, 'max_depth': 3, 'n_estimators': 200}


These are the best parameter to work with to get decent accuracy

In [59]:
best_model = grid_search.best_estimator_
accuracy = best_model.score(x_test, y_test)
print("Test Accuracy:", accuracy)

Test Accuracy: 0.7811719957209081


Here the final accuracy of our model is 78%